[DDP] Enable delayed all reduce in DDP to deprecate DelayAllReduceDDP

GD06 opened the pull request:
Summary: Enable the functionality of delaying all reduce in DDP to deprecate DelayAllReduceDDP.

Test Plan:
# Unit Test

```
buck2 test mode/opt //hpc/torchrec/tests:data_parallel_test
```

Differential Revision: D43586898

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -528,6 +528,11 @@ class DistributedDataParallel(Module, Joinable):
                          >>> ddp_logging_data = model_DDP._get_ddp_logging_data()
                          >>> static_graph = ddp_logging_data.get("can_set_static_graph")
+        delay_all_reduce_named_params (list of tuple of str and torch.nn.Parameter): a list
+                    of named parameters whose all reduce will be delayed into the backward
```
nit: will be delayed when the gradient of the parameter specified in ``param_to_hook_all_reduce`` is ready

zhaojuanmao replied:
nit: could we explain a little bit about how the two arguments will work with other arguments?

GD06 replied:
Thanks Yanli for suggestions! I revised comments to improve these.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -661,6 +684,20 @@ def __init__(
         )
 
+        # Initialize gradient buffers and register all reduce hook
+        self._grad_buffer = None
+        self._grad_views = []
```
nit: self._delay_grad_buffer, self._delay_grad_views

GD06 replied:
Done.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -661,6 +684,20 @@ def __init__(
+        # Initialize gradient buffers and register all reduce hook
+        self._grad_buffer = None
+        self._grad_views = []
+        self._use_vanilla_module = False
```
nit: self._delay_all_reduce = False

GD06 replied:
The flag "use_vanilla_module" is used to flag whether all parameters are specified in delay_all_reduce_named_params. In this case, we don't need DDP and use the forward function of the self.module.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -661,6 +684,20 @@ def __init__(
+                param_to_hook_all_reduce=param_to_hook_all_reduce,
+                device_ids=device_ids,
+            )
+            if self._use_vanilla_module:
+                return
```
nit: seems this is not needed?

GD06 replied:
When we are registering delay all reduce hook, we will check whether all parameters are specified in delay_all_reduce_named_params. If yes, we don't need DDP to construct reducer or other facilities.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -691,6 +728,52 @@ def __init__(
+
+        def _delayed_all_reduce(grad):
+            self._grad_buffer.div_(world_size)
+            _ = dist.all_reduce(self._grad_buffer, group=process_group, async_op=True)
```
where to wait for the all_reduce to complete? are you assuming optimizers will just wait for their grads?

GD06 replied:
In our use case, we assume this happens before the final all reduce (DenseArch all reduce). When they are using the same PG, we don't have to explicitly wait for this all reduce. 

But I agree that this brings the risk of missing synchronization. Do you think I should comment here about our assumptions?

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -691,6 +728,52 @@ def __init__(
+            offset = offset + param.numel()
+
+        # 5. Check whether the all reduce of all params requiring grad is delayed.
+        for module_name, module in self.module.named_modules():
+            for param_name, param in module.named_parameters(recurse=False):
+                if param.requires_grad:
+                    full_name = f"{module_name}.{param_name}"
+                    if full_name not in self.parameters_to_ignore:
+                        return
```
seems that it is not checking anything? it just simply return?

GD06 replied:
If we have param that is not specified in parameters_to_ignore, we should not set _use_vanilla_module to True.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -1190,6 +1284,9 @@ def forward(self, *inputs, **kwargs):
             )
+
+        # At the end of the forward pass, reset the grad buffer and grad views
+        self._clear_grad_buffer()
```
this can be removed?

GD06 replied:
We need to manually set grad buffer of these delayed all reduce to zero, and rebuild grad views if optimizer.zero_grad sets them to None.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -1085,10 +1168,21 @@ def _run_ddp_forward(self, *inputs, **kwargs):
         with torch.autograd.profiler.record_function(
             "DistributedDataParallel.forward"
         ):
+            if self._use_vanilla_module:
+                output = self.module.forward(*inputs, **kwargs)
+                self._clear_grad_buffer()
+                return output
```
how about the parameters that are not managed by _delay_all_reduce_params? they should be managed by DDP reducer, for these parameters, we still need to execute the below codes for preparing DDP reducer?

GD06 replied:
Yes. In the case where there are still params that are not managed by _delay_all_reduce_params, the flag _use_vanilla_module is False.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -1085,10 +1168,21 @@ def _run_ddp_forward(self, *inputs, **kwargs):
                 return self.module(*inputs, **kwargs)
 
+    def _clear_grad_buffer(self):
+        if self._grad_buffer is not None:
+            for index, param in enumerate(self._delay_all_reduce_params):
+                param.grad = self._grad_views[index]
+            self._grad_buffer.zero_()
```
just want to have a head up or you can add a comment here: "making param.grad points to the grad buffers before backward" is based on the assumption that the grad accumulation is done in place in autograd engine, for some edge cases, if the grad accumulation in autograd engine is not in place, then the param.grad and grad buffers are detached

GD06 replied:
I added a comment here to clarify this implicit assumption.

zhaojuanmao reviewed (commented):
Thanks for adding this! overall looks good, left some comments

zhaojuanmao reviewed (commented):
thanks for explaining, still some minor comments

GD06 commented on torch/nn/parallel/distributed.py:
```diff
@@ -1085,10 +1170,25 @@ def _run_ddp_forward(self, *inputs, **kwargs):
+        if self._delay_grad_buffer is not None:
+            for index, param in enumerate(self._delay_all_reduce_params):
+                param.grad = self._delay_grad_views[index]
+            self._delay_grad_buffer.zero_()
```
I changed the variable "_use_vanilla_module" to "_dar_all_params" (dar: delay-all-reduce) and added some comments to clarify code logics.

Also, thank you for your suggestions about grad_buffer! As zero_grad is a part of optimizer, we only need to rebuild param.grad when they are set to None. Thus I revised this function as follows:

We check whether the grad of all params is set to None. If yes, we will clear the whole grad buffer. If no, we will clear the grad of params individually. When iterating through self._delay_all_reduce_params, we only reset param.grad whose value is None.

zhaojuanmao commented:
overall lgtm, thank you! seems lint runner has some errors, will approve once the CI is green

GD06 commented:
/easycla

GD06 commented:
/easycla

GD06 commented:
@zhaojuanmao  Hi Yanli, CI on GitHub turns to green. Could you please take another look?

zhaojuanmao reviewed (approved):
looks good, thanks!!

GD06 commented on torch/nn/parallel/distributed.py:
```diff
@@ -661,6 +686,20 @@ def __init__(
+        # Initialize gradient buffers and register all reduce hook
+        self._delay_grad_buffer = None
+        self._delay_grad_views = []
+        self._dar_all_params = False
```
Done.

GD06 committed:
[DDP] Enable delayed all reduce in DDP to deprecate DelayAllReduceDDP (#95669)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/95669

Enable the functionality of delaying all reduce in DDP to specify the parameters whose all reduce will be hooked to a specific param. This prevents AllReduce blocking All2All in some recommendation models.

Reviewed By: zhaojuanmao

Differential Revision: D43586898

fbshipit-source-id: 1b29db7e5f15f1ee4461205ba27203dee66566fe

zhaojuanmao commented:
@pytorchbot merge

GD06 commented:
Abandon this pull request as the corresponding Diff has been abandoned.
